### ISSUE TITLE
Error building from sources.

### ISSUE THREAD

**Issue by luckydenis** [aio-libs/aiohttp#4810]
🐞 **Describe the bug**

When building according to the [instructions](https://docs.aiohttp.org/en/stable/contributing.html), an error occurs.

💡 **To Reproduce**

Correct code build.

📋 **Logs/tracebacks**

```python-traceback (paste your traceback in the next line)
...
...
...
Requirement already satisfied, skipping upgrade: PyJWT[crypto] in ./venv/lib/python3.8/site-packages (from gidgethub->cherry_picker==1.3.2->-r requirements/dev.txt (line 3)) (1.7.1)
Installing collected packages: aiohttp
  Attempting uninstall: aiohttp
    Found existing installation: aiohttp 3.6.2
    Uninstalling aiohttp-3.6.2:
      Successfully uninstalled aiohttp-3.6.2
  Running setup.py develop for aiohttp
    ERROR: Command errored out with exit status 1:
     command: /home/denis/Project/aiohttp/venv/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/home/denis/Project/aiohttp/setup.py'"'"'; __file__='"'"'/home/denis/Project/aiohttp/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps
         cwd: /home/denis/Project/aiohttp/
    Complete output (29 lines):
    **********************
    * Accellerated build *
    **********************
    running develop
    running egg_info
    writing aiohttp.egg-info/PKG-INFO
    writing dependency_links to aiohttp.egg-info/dependency_links.txt
    writing requirements to aiohttp.egg-info/requires.txt
    writing top-level names to aiohttp.egg-info/top_level.txt
    reading manifest file 'aiohttp.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    warning: no files found matching 'aiohttp' anywhere in distribution
    warning: no previously-included files matching '*.pyc' found anywhere in distribution
    warning: no previously-included files matching '*.pyd' found anywhere in distribution
    warning: no previously-included files matching '*.so' found anywhere in distribution
    warning: no previously-included files matching '*.lib' found anywhere in distribution
    warning: no previously-included files matching '*.dll' found anywhere in distribution
    warning: no previously-included files matching '*.a' found anywhere in distribution
    warning: no previously-included files matching '*.obj' found anywhere in distribution
    warning: no previously-included files found matching 'aiohttp/*.html'
    no previously-included directories found matching 'docs/_build'
    writing manifest file 'aiohttp.egg-info/SOURCES.txt'
    running build_ext
    building 'aiohttp._websocket' extension
    x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/home/denis/Project/aiohttp/venv/include -I/usr/include/python3.8 -c aiohttp/_websocket.c -o build/temp.linux-x86_64-3.8/aiohttp/_websocket.o
    x86_64-linux-gnu-gcc: error: aiohttp/_websocket.c: Нет такого файла или каталога
    x86_64-linux-gnu-gcc: fatal error: no input files
    compilation terminated.
    error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
    ----------------------------------------
  Rolling back uninstall of aiohttp
  Moving to /home/denis/Project/aiohttp/venv/lib/python3.8/site-packages/aiohttp-3.6.2.dist-info/
   from /home/denis/Project/aiohttp/venv/lib/python3.8/site-packages/~iohttp-3.6.2.dist-info
  Moving to /home/denis/Project/aiohttp/venv/lib/python3.8/site-packages/aiohttp/
   from /home/denis/Project/aiohttp/venv/lib/python3.8/site-packages/~iohttp
ERROR: Command errored out with exit status 1: /home/denis/Project/aiohttp/venv/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/home/denis/Project/aiohttp/setup.py'"'"'; __file__='"'"'/home/denis/Project/aiohttp/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps Check the logs for full command output.
```

📋 **Your version of the Python**

```console
$ python --version
Python 3.8.2
```

📋 **Your version of the aiohttp/yarl/multidict distributions**

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.6.2
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: Nikolay Kim
Author-email: fafhrd91@gmail.com
License: Apache 2
Location: /home/denis/Project/aiohttp/venv/lib/python3.8/site-packages
Requires: attrs, async-timeout, chardet, multidict, yarl
Required-by: pytest-aiohttp
```
```console
$ python -m pip show multidict
Name: multidict
Version: 4.7.6
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /home/denis/Project/aiohttp/venv/lib/python3.8/site-packages
Requires: 
Required-by: yarl, aiohttp
```
```console
$ python -m pip show yarl
Name: yarl
Version: 1.4.2
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl/
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /home/denis/Project/aiohttp/venv/lib/python3.8/site-packages
Requires: idna, multidict
Required-by: aiohttp
```

📋 **Additional context**

OS: Ubuntu 20.04 LTS

**Comment by guusbertens** [comment]
I get the same. I'm on Debian Testing (bullseye). In particular these steps give what LuckyDenis also sees:
```
git clone $REPO $DIR --recurse-submodules
cd $DIR
python3 -m venv venv
. venv/bin/activate
pip install -r requirements/dev.txt
```
These are the suggested steps from the [contributing section](https://docs.aiohttp.org/en/stable/contributing.html). However I need to add the following steps before the last step:
```
pip install wheel
pip install cython
make cythonize
```
With this everything works as expected.

**Comment by guusbertens** [comment]
Indeed, setup.py states that the makefile runs ```cythonize```. The [setuptools documentation](https://setuptools.readthedocs.io/en/latest/setuptools.html#distributing-extensions-compiled-with-cython) confirms that setuptools _doesn't_ run cythonize; it only compiles .c files, and hence the cythonized code should be included in the git repo. The cython documentation [suggests](https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#multiple-cython-files-in-a-package) using the ```cythonize``` function from within setup.py, so that pip can cythonize all that is needed.

**Comment by luckydenis** [comment]
It seems that this point is not obvious and it would be good to reflect this in the documentation or add the necessary commands to the installation scripts.
@webknjaz, @guusbertens

**Comment by webknjaz** [comment]
We moved cythonize out of setup.py in order to have Cython an optional build dependency.

**Comment by webknjaz** [comment]
I don't see you running `make test`. What exactly did you do?

**Comment by guusbertens** [comment]
@webknjaz I just ran this:

```
git clone $REPO $DIR --recurse-submodules
cd $DIR
python3 -m venv venv
. venv/bin/activate
pip install wheel
pip install cython
make cythonize
pip install -r requirements/dev.txt
make test
```

This works like a charm.

Do you agree that, given what is said in the setuptools docs and the Cython docs, the current breakage is expected, and that this issue is indeed a bug, and isn't an invalid issue?

You mention that you keeping Cython optional is important. As far as setup.py is concerned, Cython could be in there while being optional, *but* optional dependencies aren't supported by pyproject.toml, and this causes the problem. Am I right about this? If this is the case, then I think there are two options:

* include cythonize'd C-files in the git repo; or
* update the documentation to add the cythonize step; or
* maybe using two different pyproject.toml files, but I think this overcomplicates matters a lot...

Am I missing something?

**Comment by webknjaz** [comment]
> Do you agree that, given what is said in the setuptools docs and the Cython docs, the current breakage is expected, and that this issue is indeed a bug, and isn't an invalid issue?

This is controlled by `AIOHTTP_NO_EXTENSIONS=1` env var. If it's set, `setup.py` won't attempt to build the C-extensions. sdists on PyPI are shipped with pre-cythonized sources (C-files) and during development you really have to be explicit about what you want and also the default is to build the extensions (so that you accidentally don't test pure-python while thinking that you also test Cython).

> You mention that you keeping Cython optional is important. As far as setup.py is concerned, Cython could be in there while being optional, _but_ optional dependencies aren't supported by pyproject.toml, and this causes the problem. Am I right about this? If this is the case, then I think there are two options:
> 
> * include cythonize'd C-files in the git repo; or
> * update the documentation to add the cythonize step; or
> * maybe using two different pyproject.toml files, but I think this overcomplicates matters a lot...
> 
> Am I missing something?

Before `pyproject.toml` was a thing and before pip started doing mandatory build isolation we could ask users to `pip install Cython` into the same env and that would do the trick. Now we cannot do this because the build env is always different. And if we put `Cython` into `setup_requires`, it'd be mandatory. We cannot do this because even just installing Cython often requires GCC (at least it used to).
Oh, and most of the tools don't have support for having a different `pyproject.toml` source.

Now, if you look at the instructions, there's `make test` which, if you inspect `Makefile` has a chain of dependencies `test -> .develop -> .install-deps -> cythonize -> "running cython over all sources"`.

So just running `make test` _should_ also cythonize stuff for development. If you know how to improve the docs, though, I'm always open to reasonable improvements.

**Comment by guusbertens** [comment]
@webknjaz Thanks for your extensive answer. I agree, there is ```make test```, but if one follows the instructions on the [Contributing](https://docs.aiohttp.org/en/stable/contributing.html) page, one will try to take these steps:

1. git clone
2. create and activate venv
3. pip install -r requirements/dev.txt
4. make test

Step 3 fails - you wouldn't make it to step 4. However, as you mention, due to the way the Makefile is set up, it will do the right thing even without step 3 ever being done.

In short, I think it's best to just remove ```pip install -r requirements/dev.txt``` from the contributing page, and have people go directly to ```make test```.

Also, on that same page: it seems like ```make cov``` doesn't work. If I install ```tox``` it works, but it doesn't give the expected result. The command ```make cov-dev``` works immediately, and gives the expected result. Should I open a new issue for that?

**Comment by webknjaz** [comment]
1. Yes, I guess that could be removed.
2. tox is not integrated and is likely out of date. You may send a PR right away.
     If you look at the CI, it doesn't even use make for that anymore. IMHO there's
     a lot of things that could be refactored, I'd even replace make with tox but
     some other maintainers don't share this opinion so it's stuck in this half-
     broken state now.

**Comment by guusbertens** [comment]
Okay, I'll prepare a PR for both things later today. Thanks for your feedback!

**Comment by guusbertens** [comment]
Hah, #4161 already solves most of this. However, that documentation didn't go to docs.aiohttp.org yet, which is why I missed it.

**Comment by jmgurney** [comment]
I'd like to see `setup.py` detect that the C files are missing, and automatically set `AIOHTTP_NO_EXTENSIONS=1`.  W/o this, it means that it's not really possible to build from source as a dependency from another package.

something like:
```
diff --git a/setup.py b/setup.py
index 2622e5c5..33e8e6fe 100644
--- a/setup.py
+++ b/setup.py
@@ -44,6 +44,8 @@ extensions = [
     Extension("aiohttp._http_writer", ["aiohttp/_http_writer.c"]),
 ]
 
+if not all((pathlib.Path(fname).exists() for ext in extensions for fname in ext.sources)):
+    NO_EXTENSIONS = True
 
 build_type = "Pure" if NO_EXTENSIONS else "Accelerated"
 setup_kwargs = {} if NO_EXTENSIONS else {"ext_modules": extensions}

```

**Comment by webknjaz** [comment]
Absolutely not. This would mean that people unknowingly get a slow implementation by default, which is unaccessible. The proper solution is to use dists on PyPI that bundle everything necessary.
